### ISSUE TITLE
[HOLD for payment 2025-01-18] [$250] Search Parser does not recognize quotation marks on iOS

### ISSUE THREAD

**Issue by 289Adam289** [Expensify/App#54192]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:**
**Slack conversation** (hyperlinked to channel name):

## Action Performed:
1. Open Search Router
2. Input `“a b”` using iOS keyboard and press enter
3. Notice that Search Router does not work properly.

## Expected Result:
Describe what you think should've happened

`"a b"` should be displayed. Parser should recognize `”` quotation mark on iOS.

## Actual Result:
Describe what actually happened

It displays `""“a"" "" b”""`, Parser did not recognize `”`.
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [x] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

https://github.com/user-attachments/assets/2a19d115-1f06-4172-aa2b-775429d4cdea

  
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021868706886680875349</li>
        <li>Upwork Job ID: 1868706886680875349</li>
        <li>Last Price Increase: 2024-12-16</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @joekaufmanexpensify</details>

**Comment by joekaufmanexpensify** [comment]
I can reproduce this. Not sure if this came from another issue or Slack thread, so will wait until @luacmartins chimes in on the context for this.

**Comment by luacmartins** [comment]
All yours @289Adam289

**Comment by JoshIri360** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Search Parser does not recognize quotation marks on iOS

### What is the root cause of that problem?
The root cause is that iOS keyboards use different Unicode characters for quotation marks (curly quotes: U+201C and U+201D) compared to the straight quotes (U+0022) that our search parser expects. The current code doesn't normalize these different types of quotation marks before passing the query to the parser, leading to incorrect parsing and query malformation.

### What changes do you think we should make in order to solve the problem?
In `src/libs/SearchQueryUtils.ts`:

```typescript
function buildSearchQueryJSON(query: SearchQueryString) {
    // First normalize all types of quotes (straight, left curly, right curly) to straight quotes
    const sanitizedQuery = query
        .replace(/[\u201C\u201D\u0022]/g, '"')  // Unicode for ", ", "
        .replace(/"{2,}/g, '"')                  // Replace multiple quotes with single
        .trim();

    try {
        const result = searchParser.parse(sanitizedQuery) as SearchQueryJSON;
        const flatFilters = getFilters(result);

        // Add the full input and hash to the results
        result.inputQuery = query;
        result.flatFilters = flatFilters;
        const {primaryHash, recentSearchHash} = getQueryHashes(result);
        result.hash = primaryHash;
        result.recentSearchHash = recentSearchHash;

        return result;
    } catch (e) {
        console.error(`Error when parsing SearchQuery: "${sanitizedQuery}"`, e);
    }
}
```

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?
Test quote normalization with straight quotes, curly quotes, multiple quotes, and mixed quotes in search queries. Include tests for iOS keyboard input simulation and verify proper parsing of quoted phrases with spaces.

### What alternative solutions did you explore? (Optional)
An important note: The current desktop behavior searches for `""a a""` (with explicit quotes), not `"a a"` as intended. We could alternatively remove quotations entirely to match the intended search behavior, but this would be a more significant change to the search functionality. The proposed solution maintains consistency with the current desktop behavior while fixing the iOS-specific issues.

If you want to modify this behavior, you could add the space character to the regexp character class like this:
```typescript:src/libs/SearchQueryUtils.ts
const regexp = /[^A-Za-z0-9_@./#&+\-\\';," ]/g;
```

However, be careful about modifying this as it might affect how the search query parser interprets the terms.

**Comment by joekaufmanexpensify** [comment]
@JoshIri360 This issue isn't actively looking for volunteers. You can find issues that need volunteers/proposals by filtering for those with the `Help Wanted` label applied. Let me know if you have any questions. Thanks!

**Comment by JoshIri360** [comment]
@joekaufmanexpensify 
Ah, that's my bad, thanks for clearing things up.

**Comment by joekaufmanexpensify** [comment]
Course!

**Comment by joekaufmanexpensify** [comment]
@289Adam289 is there an ETA for a PR here?

**Comment by 289Adam289** [comment]
Hi, sorry I got caught up in Search highlighting task. I should have draft ready today.

**Comment by joekaufmanexpensify** [comment]
As an fyi, I'm going to be OOO after today through the new year. If anything urgently is needed from a BZ member during that time, please reach out in slack. Otherwise, I will handle any payment that becomes required here when I return!

**Comment by luacmartins** [comment]
We're prioritizing search highlighting instead. We'll get back to this soon.

**Comment by luacmartins** [comment]
Update above ^

**Comment by joekaufmanexpensify** [comment]
Sounds good! I think this is still pending PR once the highlight issue mentioned above is done.

**Comment by joekaufmanexpensify** [comment]
This is a pretty niche bug, so I don't think we need a specific regression test for it every time. Checklist otherwise looks great!

**Comment by joekaufmanexpensify** [comment]
@fedirjh please request payment at your earliest convenience. Closing the issue as this otherwise all set!

**Comment by garrettmknight** [comment]
$250 approved for @fedirjh
